### PR TITLE
Tests: refactor callbacks

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?
Fixes # (issue).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@Borda are you sure it is good to replace all model definitions with the same one? Some tests may be based on the fact that only a subset of functionality is provided.

File: tests/callbacks/test_callbacks.py [hunk 1]
```diff
@@ -214,18 +200,12 @@ def on_test_end(self, trainer, pl_module):
...
-            output = super().training_step(*args, **kwargs)
-            loss = output['loss']  # could be anything else
-            output.update({'my_train_metric': loss})
-            return output
-
-    model = ModelWithoutValStep(tutils.get_default_hparams())
 
-    stopping = EarlyStopping(monitor='my_train_metric', min_delta=0.1)
+    model = EvalModelTemplate(tutils.get_default_hparams())
+    model.validation_step = None
```
Comment (jeremyjordan) [inline tests/callbacks/test_callbacks.py:205]: do we want this to be `None` or do we want `model.validation_step = LightningModule.validation_step`?

Comment (Borda) [reply tests/callbacks/test_callbacks.py:205]: I had it also originally with `LightningModule.validation_step` but @williamFalcon commented that he prefers none... For me `LightningModule.validation_step` would be easier to understand that it is not implemented...

**Comment by Borda** [comment]
> @Borda are you sure it is good to replace all model definitions with the same one? Some tests may be based on the fact that only a subset of functionality is provided.

the missing subset is implemented as you have the full model and mak only the missing so pls point out where I did not mask it properly...

File: tests/callbacks/test_callbacks.py [hunk 2]
```diff
@@ -214,18 +200,12 @@ def on_test_end(self, trainer, pl_module):
 
 def test_early_stopping_no_val_step(tmpdir):
     """Test that early stopping callback falls back to training metrics when no validation defined."""
-    class ModelWithoutValStep(LightTrainDataloader, TestModelBase):
-
-        def training_step(self, *args, **kwargs):
-            output = super().training_step(*args, **kwargs)
-            loss = output['loss']  # could be anything else
-            output.update({'my_train_metric': loss})
```
Comment (awaelchli) [inline tests/callbacks/test_callbacks.py:222]: For example here. This is not just a refactoring, this is changing the intention of the test. By adding a custom key we test that early stopping works on the given key and not just the default loss key. If you reduce all tests to the same template model, we lose the testing of the edge cases. Please consider keeping the train step here.

File: tests/callbacks/test_callbacks.py [hunk 3]
```diff
@@ -214,18 +200,18 @@ def on_test_end(self, trainer, pl_module):
...
         def training_step(self, *args, **kwargs):
             output = super().training_step(*args, **kwargs)
-            loss = output['loss']  # could be anything else
-            output.update({'my_train_metric': loss})
+            output.update({'my_train_metric': output['loss']})  # could be anything else
             return output
 
-    model = ModelWithoutValStep(tutils.get_default_hparams())
+    model = CurrentModel(tutils.get_default_hparams())
+    model.validation_step = LightningModule.validation_step
```
Comment (williamFalcon) [inline tests/callbacks/test_callbacks.py:211]: Again, no need to use LightningModule...

Comment (Borda) [reply tests/callbacks/test_callbacks.py:211]: we found it cleaner to see that this method was not overwritten... so you want to have it?
```
model.validation_step = None
```
